vars/folioHelm.groovy

```diff
@@ -128,31 +129,114 @@ void checkPodRunning(String ns, String podName) {
+//        try {
+//          evictedPodsList = sh(script: "kubectl delete pod -n ${ns} --field-selector=\"status.phase==Failed\"", returnStdout: true)
+//        } catch (Error err) {
+//          new Logger(this, "managePods").warning("Error: " + err.getMessage() + "\nList of evicted pods: ${evictedPodsList}")
+//        }
+//      } else {
+//        println('All pods are running.')
+//      }
+//    }
+//  }
+//}
+
+void checkAllDeploymentsRunning(String ns, def deploymentsInput) {
+  println('Starting deployment monitoring...')
+
+  // Check the input parameter and convert it to a list
+  List<FolioModule> deploymentsList
+  if (deploymentsInput instanceof List) {
+    deploymentsList = deploymentsInput
+  } else if (deploymentsInput instanceof FolioModule) {
+    deploymentsList = [deploymentsInput]
+  }
 
-      def result = sh(script: "kubectl get pods -n ${ns} --no-headers | awk '{print \$3}'", returnStdout: true).trim()
+  boolean allDeploymentsUpdated = false
+  int timer = 0
+  int maxTime = 10 * 60 // 10 minutes in seconds
+
+  try {
+    while (!allDeploymentsUpdated) {
+      def jsonOutput
+      try {
+        // Execute the kubectl command
+        jsonOutput = sh(
+          script: "kubectl get deployments -n ${ns} -o json",
+          returnStdout: true
+        ).trim()
+      } catch (Exception e) {
+        error("Failed to execute kubectl command: ${e.message}")
+      }
+
+      def deploymentsJson
+      try {
+        // Parse the JSON output
+        deploymentsJson = readJSON text: jsonOutput
+      } catch (Exception e) {
+        error("Failed to parse JSON output: ${e.message}")
+      }
 
-      notAllRunning = result.split('\n').any { status -> status != 'Running' }
+      // Check if there are any deployments in the namespace
+      if (!deploymentsJson.items || deploymentsJson.items.isEmpty()) {
+        error("No deployments found in namespace '${ns}'. Please check the namespace or deployment configuration.")
+      }
 
-      if (notAllRunning) {
-        def evictedPodsList
-        println('Not all pods are running. Retrying...')
-        try {
-          evictedPodsList = sh(script: "kubectl delete pod -n ${ns} --field-selector=\"status.phase==Failed\"", returnStdout: true)
-        } catch (Error err) {
-          new Logger(this, "managePods").warning("Error: " + err.getMessage() + "\nList of evicted pods: ${evictedPodsList}")
+      def unfinishedDeployments = []
+
+      // Check each deployment from the list
+      deploymentsList.each { folioModule ->
+        def deployment = deploymentsJson.items.find { it.metadata.name == folioModule.name }
+        if (deployment) {
+          if (deployment.status.updatedReplicas != deployment.spec.replicas ||
+            deployment.status.readyReplicas != deployment.spec.replicas ||
+            deployment.status.unavailableReplicas > 0 ||
+            deployment.status.conditions.any { it.type == "Available" && it.status == "False" }) {
```

Comment:
deployment.status and deployment.spec.replicas is accessed multiple times. This could be extracted into a variable to reduce repetition and improve clarity.

Reply:
done